### ISSUE TITLE
Fixed size ModelArray/Partitioning

### ISSUE THREAD

**Issue by athelaf** [nextsimhub/nextsimdg#110]
Having a fixed-size ModelArray that can only be resized when we read the input file implies that to distribute the model to multiple processes we also need to know the partitioning ahead of time, which means we need to do the partitioning offline. So I assume this is the approach we want to take, i.e., do the domain decomposition separately using for example the domain decomposition tool I have written with Zoltan and save the partitions in a file that will then be read during model initialization?

**Comment by einola** [comment]
In my opinion, we should do partitioning offline. We hadn't made a formal decision, but if you (@athelaf and @timspainNERSC) don't find a serious flaw in my logic below, then we'll go for that approach.

I think doing the partitioning offline, as a preprocessing step, is preferable for two main reasons.

Pros of offline preprocessing:
 - Separation of processing or "do one thing and do it well".
 - Keeping partitioning offline makes it easier to maintain or replace the partitioning software

Cons:
 - Heavier and more complex pre-processing
 - We'd need to re-process a restart/init file if we want to change the partitioning.

Am I missing something?

For me, it's only really the last "con" that is of any importance. But I think it should be ok still.

**Comment by timspainNERSC** [comment]
If the partitioning is provided in one of the config files, then I don't think the last "con" applies either.
The workflow I envisage for this would be:

1. The pre-processing run is run on the restart file for the current architecture, and spits out a 'partition.cfg' (or something similar).
2. This is then read in as one of the config files on the command line, and the partition configuration applied in the model set up
3. Any further required magic also occurs (spawning other processes, whatever)

Is this a possible workflow?
The advantage here is that the partitioning is not baked into the restart file, which I feel is better reserved for model data only. Any local configuration is then in the config files, which have to be shared in any case. Another user could use the same restart netCDF file on their own machine, and if necessary create their own partitioning config.

In the case of no partitioning config files, then it runs as a single node (?) process.

**Comment by einola** [comment]
Yes, this is similar to the workflow I had in mind. Not the same, but that's below. The "con" here is that you have to do pre-processing on the restart file in the edge case that you want to restart, but with a different partitioning. In models I've used, you just give the config and restart files to the model and it sorts everything out - including the partition map. So, pre-processing is an extra step.

Anyway, what I had in mind was that the partition map would be stored in the restart file. I didn't think we could store it in a config file, but maybe that's wrong(?). The options for running the model are then:

 1. A "fresh start": Use a pre-processing script to read in data from various sources to create a "restart" (or init) file. This script calls Zoltan (or something else) to do the partitioning. The model reads the restart/init file and config file(s).
 2. An "ordinary" restart: We don't want to change anything in the setup, just continue the simulation from where it last stopped. Use a restart file and the config file(s) from the last run (the config file could be optional if all the options are stored in the restart file).
 3. A restart with modified options: Use the restart file like before, but load a modified config file.
 4. A restart with a different partitioning: Pre-process the restart file with a script that calls Zoltan. Then use the new restart file and a config file (optional again, or maybe modified).

Here 4. is the edge case ... but I don't think it's very important. It could become a stupid "gotcha" if people restart, but change the number of MPI processes without pre-processing the restart file. That would leave them wondering why more nodes don't lead to better performance. But this we can surely catch - right?

**Comment by timspainNERSC** [comment]
The pre-processing could be run as part of the model, if necessary. Perhaps check the partition parameters against the configured number of nodes, and if there is any inconsistency, launch the pre-processing process before restarting the model proper with the new configuration.

**Comment by timspainNERSC** [comment]
I've created a new issue that was inspired by this, #113, on how the configuration data moves into and out of the model.

**Comment by einola** [comment]
Ok, good. I've commented on #113 too.

We agree then, that partitioning is done in pre-processing?

**Comment by athelaf** [comment]
I agree that the partitioning should be de-coupled from nextdim_dg and done offline, as this allows us to use multiple partitioning libraries without affecting the nextsim_dg codebase and introducing dependencies. 

Is the land-sea mask part of the restart/model file? For partitioning we only need the mask, so I believe it should be quite simple to process this file in a pre-processing step and spit out a partition.cfg file as @timspainNERSC suggests. If the user for some reason tries to use a number of processes that does not match the partition.cfg file, the simplest thing we can do is just generate an error and ask them to run the pre-processing step for the number of processes they want.

**Comment by timspainNERSC** [comment]
The land-sea mask is in the restart file, if it exists. No `mask` field in the file means every point is ocean.

The file `rect3030.res.nc` in the branch `issue100_output` contains an example of a land mask.

**Comment by athelaf** [comment]
Excellent, I am refining and adapting the partitioning tool to reflect the restart file format.

**Comment by TomMelt** [comment]
I believe this can be closed now. Our workflow is to run decomp tool first, generating the metadata offline. Then we read it when running nextsim.
